flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/MergeTableLikeUtil.java

```diff
@@ -312,113 +302,36 @@ private void appendDerivedPrimaryKey(@Nullable SqlTableConstraint derivedPrimary
-            }
-
-            List<String> components = eventTimeColumnName.names;
-            if (!allFieldsTypes.containsKey(components.get(0))) {
-                throw new ValidationException(
-                        String.format(
-                                "The rowtime attribute field '%s' is not defined in the table schema, at %s\n"
-                                        + "Available fields: [%s]",
-                                fullRowtimeExpression,
-                                eventTimeColumnName.getParserPosition(),
-                                allFieldsTypes.keySet().stream()
-                                        .collect(Collectors.joining("', '", "'", "'"))));
-            }
-
-            if (components.size() > 1) {
-                RelDataType componentType = allFieldsTypes.get(components.get(0));
-                for (int i = 1; i < components.size(); i++) {
-                    RelDataTypeField field = componentType.getField(components.get(i), true, false);
-                    if (field == null) {
+            if (mergingStrategies.get(SqlTableLike.FeatureOption.WATERMARKS)
+                    != SqlTableLike.MergingStrategy.OVERWRITING) {
+                for (SqlWatermark derivedWatermarkSpec : derivedWatermarkSpecs) {
+                    SqlIdentifier eventTimeColumnName =
+                            derivedWatermarkSpec.getEventTimeColumnName();
+                    String rowtimeAttribute = eventTimeColumnName.toString();
+
+                    if (watermarkSpecs.containsKey(rowtimeAttribute)) {
                         throw new ValidationException(
                                 String.format(
-                                        "The rowtime attribute field '%s' is not defined in the table schema, at %s\n"
-                                                + "Nested field '%s' was not found in a composite type: %s.",
-                                        fullRowtimeExpression,
-                                        eventTimeColumnName.getComponent(i).getParserPosition(),
-                                        components.get(i),
-                                        FlinkTypeFactory.toLogicalType(
-                                                allFieldsTypes.get(components.get(0)))));
+                                        "There already exists a watermark spec for column '%s' in the base table. You "
+                                                + "might want to specify EXCLUDING WATERMARKS or OVERWRITING WATERMARKS.",
+                                        rowtimeAttribute));
                     }
-                    componentType = field.getType();
                 }
             }
+
+            HashMap<String, RelDataType> nameToTypeMap = new LinkedHashMap<>();
+            nameToTypeMap.putAll(physicalFieldNamesToTypes);
+            nameToTypeMap.putAll(metadataFieldNamesToTypes);
+            nameToTypeMap.putAll(computedFieldNamesToTypes);
```

Comment:
I left this like before without adding them to a global Map. I noticed the LIKE table option does not use the compute fields when evaluating other compute columns. But metadata do evaluate the expressions from compute pools. I don't know if that's allowed, or if we can use other compute columns to evaluate compute expressions?